### ISSUE TITLE
seq-to-seq ASR batch Scheduler with multiple datasets

### ISSUE THREAD

**Issue by GaelleLaperriere** [speechbrain/speechbrain#650]
Hello,

I am using the **sequence-to-sequence ASR system with CommonVoice** recipe. I made a many-to-many and one-to-many systems to use 2 different datasets (one is with Italian data and the other with French data). 

To do so, I used the `RandomSampler` and `ConcatDatasetBatchSampler` from `speechbrain.dataio.sampler` to return `DataLoader` with the `dataio_prepare` function. I am trying to implement a scheduler to switch between tasks (Italian and French) following a probability. For now, I can alternate between Italian and French batches but they are paired due to the `ConcatDatasetBatchSampler`. 
On top of that, I have more French than Italian data. For now, I only see the same amount of French samples that I have in my Italian dataset. I need to see all of both datasets. The idea I had (below) could allow me to do so. 

I don't want to just use one of the two batches paired (French or Italian) for one iteration, following my probability. Doing so will make an epoch see only half my data. Even if all data will be seen thanks to X epochs and a shuffle, the training losses would not be analyzable between epochs. I'm searching for a way to feed all of it to the NN in one epoch, choosing the task (French or Italian) for each pair of batches given. 

I though of a way to do so and would like your opinion: 
- I could return 2 `DataLoader` for each corpus with the `dataio_prepare()` function.
- And override the `fit()` function from the `core.py` script to get both `DataLoader` instead of one. (I'm not sure, `self.train_sampler` would stay `None`, knowing that `make_dataloader()` won't be called, right ?)
- Then, I could use [Tqdm-multiprocess](https://pypi.org/project/tqdm-multiprocess/) instead of Tqdm, to give batches (Italian OR French) to `fit_batch()`. It would allow me to choose which task to apply here without "wasting" the other task batches.

If you have another idea on how to do this, do not hesitate. 
Here is the way I am currently giving batches (exemple for the training dataset only, and waves only):

```python
def dataio_prepare(hparams):
    ...
    # Concat linear
    train_data = torch.utils.data.ConcatDataset([train_data_fr, train_data_it])
    # Make samplers
    sampler_train_fr = RandomSampler(train_data_fr)
    sampler_train_it = RandomSampler(train_data_it)
    sampler_train = ConcatDatasetBatchSampler([sampler_train_fr, sampler_train_it], [hparams["fr_batch_size"], hparams["it_batch_size"]])
    # Make DataLoader
    dataloader_train = torch.utils.data.DataLoader(train_data, batch_sampler=sampler_train, num_workers=3, collate_fn=PaddedBatch)
    return dataloader_train, dataloader_valid, ...

class ASR(sb.core.Brain):
    ...
    def fit_batch(self, batch):
        # Get all needed from batch
        batch = batch.to(self.device)
        wavs = batch.sig
        ...
        # Basic python slices
        wavs_fr, wavs_it = wavs[:fr_batch_size], wavs[it_batch_size:]
        ...
        # For French
        predictions = self.compute_forward(...)
        loss = compute_objectives(...)
        loss.backward()
        ...
        # Same for Italian
        predictions = self.compute_forward(...)
        loss = compute_objectives(...)
        loss.backward()
        ...
```

Thank You.

-------------------------------------------------------

**EDIT** : Worked fine, didn't need the Tqdm-multiprocess. I used Tqdm with a `range(nb_batches + nb_tasks - 1)` list instead of the DataLoader itself. I iterated through the different DataLoaders, with `iter(dataloader)`, with the Tqdm in the `fit()` function:

```python
task = tasks[batch_index%len(tasks)] 
try:
    batch = next(dataloader_iterator[task])
    ... # do treatments as usual
except StopIteration:
    del dataloader_iterator[task]
    tasks.remove(task)
    pass
```

This way to get the task of the next batch alternate between all the tasks evenly. By replacing `task = tasks[batch_index%len(tasks)] `, I was able to use a probability, like I wanted, and more.

**Comment by TParcollet** [comment]
@popcornell @Gastron

**Comment by popcornell** [comment]
Hi, 

Probably a more straightforward way to do this is to append fake examples to the shorter dataset (italian or french) then in the training loop you check for example` if not torch.all(input_to_your net)` if the dummy data is all zeros. In this way you can use the same pipeline you are using right now with minimal modifications.

**Comment by TParcollet** [comment]
But as a side note: we should simplify the support of many dataloaders, no ?

**Comment by GaelleLaperriere** [comment]
Yes, but this way I would need to set the sequence of tasks (French - Italian - Italian - Italian - French - ...) when creating the DataLoader, to break the "pair problem". It's the easiest way but not the cleaner. I would like to be able to change the sequence in the middle of the training, for future experiments.

**Comment by TParcollet** [comment]
I think you can use your approach, but you have to check what is dataloader related on the Brain class and override it (like the sampler).

**Comment by popcornell** [comment]
You want to train for X steps on dataset 1 and then for Y steps on dataset 2 is this correct ? 
And do this dynamically right ?
If so yes you have pretty much to override everything in the Brain class fit method and possibly even more.

**Comment by popcornell** [comment]
Well actually there may be a workaround by implementing a custom sampler where the sampling is conditioned on a variable, 
if you are able in the training loop to change that variable like self.train_loader.sampler.dataset == "french" you could probably switch the dataset dynamically

**Comment by TParcollet** [comment]
But the sampler works on a single Dataset object right?

**Comment by popcornell** [comment]
Yes but you can have a ConcatDataset and play with indexes.

**Comment by GaelleLaperriere** [comment]
> You want to train for X steps on dataset 1 and then for Y steps on dataset 2 is this correct ?
> And do this dynamically right ?

Yes, that's it. 

> Well actually there may be a workaround by implementing a custom sampler where the sampling is conditioned on a variable,
> if you are able in the training loop to change that variable like self.train_loader.sampler.dataset == "french" you could probably switch the dataset dynamically

Thanks, I will look into it. I began changing the `fit()` function. If I encounter a problem, I will try this approach.
